### ISSUE TITLE
Shortcuts: keydown-Listener not removed on client-side

### ISSUE THREAD

**Issue by Springrbua** [vaadin/flow#12307]
### Description of the bug
I have a dialog in which I register a few `Shortcut`s. For performance reasons when closing the dialog, I only hide it by setting the overlay to `display:none`, to be able to reuse it, instead of closing it.
When opening/showing the dialog, I remove all `ShortuctEventListener`s and register new ones, since the available actions could change in the meantime.
Now if I open and close the dialog mulitple times, the `ShortcutEventListener` gets triggered multiple times. While debugging, I found that the `ShortcutRegistration` instance is always the same, so it seems like there are multiple events coming from the client.
Inspecting the element on the client-side it indeed seems like the old listeners are not removed:
Dialog after opening once:
![image](https://user-images.githubusercontent.com/18258317/140744313-51eedfef-b443-4218-83d9-ad602fae14fc.png)
Dialog after opening twice:
![image](https://user-images.githubusercontent.com/18258317/140744446-f04f2236-6fe0-4d22-bcbb-f80296a6b1c9.png)

### Minimal reproducible example

1. Download a [vaadin flow starter](https://start.vaadin.com/app/) with an empty view
2. Add the following code to the `EmptyView.java`:
```
@PageTitle("Empty")
@Route(value = "empty", layout = MainLayout.class)
@RouteAlias(value = "", layout = MainLayout.class)
public class EmptyView extends VerticalLayout {
  private final Dialog dlg;
  private Registration shortcutRegistration = () -> {};

  public EmptyView() {
    dlg = new Dialog();
    dlg.setCloseOnEsc(false);
    dlg.setCloseOnOutsideClick(false);
    Shortcuts.setShortcutListenOnElement("this.$.overlay", dlg);
    VerticalLayout vlDlg = new VerticalLayout();
    vlDlg.getElement().setAttribute("tabindex", "-1");
    vlDlg.add(
      new Span("Press 'Del' to trigger shortcut"),
      new Span("Press 'Esc' to hide dialog")
    );
    dlg.add(vlDlg);
    add(new Button("Open Dialog", e -> {
      updateShortcuts();
      if (!dlg.isOpened())
        dlg.open();
      else {
        dlg.setModal(true);
        dlg.getElement().executeJs(  "this.$.overlay.style.display = \"\";");
        vlDlg.getElement().executeJs("this.focus();");
      }
    }));
  }
  private void updateShortcuts() {
    shortcutRegistration.remove();
    shortcutRegistration = Registration.combine(
      Shortcuts.addShortcutListener(this, e -> {
        dlg.setModal(false);
        dlg.getElement().executeJs("this.$.overlay.style.display = \"none\";");
      }, Key.ESCAPE).listenOn(dlg),
      Shortcuts.addShortcutListener(this, e -> System.out.println("Received del"), Key.DELETE)
        .listenOn(dlg)
    );
  }
}
```
3. Run the application and open the dialog by clicking the button
4. Close (hide) the dialog with `ESC`
5. Reopen the dialog by clicking the button
6. Press `DEL` to trigger the other shortcut and see the logs in the console.
   The `DEL` shortcut was logged multiple times

### Expected behavior
Removing the `ShortcutRegistration` should also remove the client-side listener

### Actual behavior
Removing the `ShortcutRegistration` does not remove the client-side listener and therefor the server-side listener is triggered multiple times.

### Versions:
    - Vaadin / Flow version: 21.0.3
    - Java version: 11.0.10 AdoptOpenJDK
    - OS version: Windows 10 64bit
    - Browser version (if applicable): Chrome 95.0.4638.69
    - IDE (if applicable): IntelliJ IDEA Community Edition

**Comment by Avec112** [comment]
I might experiencing a similar problem where the same trigger run both events.

I had two dialoges where the first triggered the next dialog.
The first (outer) dialog had Shortcut Key.ENTER while the next (inner) dialog had `textfield.addKeyUpListener(Key.Enter, ...)`

When first dialog opened and I hit Key.ENTER on my keyboard, both events was triggered. 

Changing to `textfield.addKeyPressListener(Key.Enter, ...)` in the second dialog solved the problem but I think this just hides a possible bug.

The key state seems to be this order: Down, Press (bottom), Up

It also seems `Shortcuts.addShortcutListener(confirmButton, confirmButton::click, Key.ENTER);` is picked up by the second event KeyUp and not handling it as a separate event.

**Comment by pleku** [comment]
The problem in [this case](https://github.com/vaadin/flow/issues/12307#issue-1047501466) is not with the shortcut not being removed, but the special case handling for `vaadin-dialog` which registers a second listener via JS to the overlay that fires a new event on the actual `vaadin-dialog` based on the original event.
This is done each time when you register a new shortcut, as it was not deemed a requirement to re-register shortcuts on **the same opened Dialog-element over and over again**. It is a bug though your code is also doing some things that in my personal opinion it should not (see points later on). I have some ugly ideas how-to fix it.

In the meanwhile, to workaround it @Springrbua, you could adopt your code to:

```java
private final Registration haxorRegistration; // 1) store this inside the dialog creation + "opening" class
...   
haxorRegistration = Shortcuts.setShortcutListenOnElement("this.$.overlay", dlg); // do this only once when creating the dialog or opening it for the first time
...
haxorRegistration.remove(); // do this call when you "close" the dialog for the first time. This way it will not be added again
```
This is a workaround, but I tested it and it works.

Two additional comments on the code example:
1) This line would be unnecessary as the `Dialog` does it already
```java
    Shortcuts.setShortcutListenOnElement("this.$.overlay", dlg);
```
in https://github.com/vaadin/flow-components/blob/master/vaadin-dialog-flow-parent/vaadin-dialog-flow/src/main/java/com/vaadin/flow/component/dialog/Dialog.java#L658
You'll need it for the workaround above though

2) You're only hiding the dialog in the DOM, which means that someone could make it visible again and keep on using it even when they should not. It would be safer to also make it disabled with `dialog.setEnabled(false)`

**Comment by pleku** [comment]
@Avec112 I'm not 100 % the problem you describe is the same issue, my guess would be that it is not.

If your issue still reproduces after the bug described in this ticket is fixed, you should create open a new issue with the code to reproduce the issue.

**Comment by Avec112** [comment]
Will do.

**Comment by Springrbua** [comment]
I just tested the workaround and it seems to work, thank you @pleku.
